lightning/src/ln/channel.rs

```diff
@@ -1488,17 +1483,21 @@ impl<ChanSigner: ChannelKeys> Channel<ChanSigner> {
 			}
 		};
 
+		// Now that we're past error-generating stuff, update our local state:
+
 		let their_pubkeys = self.their_pubkeys.as_ref().unwrap();
 		let funding_redeemscript = self.get_funding_redeemscript();
-		self.channel_monitor.set_basic_channel_info(&their_pubkeys.htlc_basepoint, &their_pubkeys.delayed_payment_basepoint, self.their_to_self_delay, funding_redeemscript.clone(), self.channel_value_satoshis, self.get_commitment_transaction_number_obscure_factor());
-
 		let funding_txo_script = funding_redeemscript.to_v0_p2wsh();
-		self.channel_monitor.set_funding_info((funding_txo, funding_txo_script));
-
-		// Now that we're past error-generating stuff, update our local state:
-
-		self.channel_monitor.provide_latest_remote_commitment_tx_info(&remote_initial_commitment_tx, Vec::new(), self.cur_remote_commitment_transaction_number, self.their_cur_commitment_point.unwrap());
-		self.channel_monitor.provide_latest_local_commitment_tx_info(local_initial_commitment_tx, local_keys, self.feerate_per_kw, Vec::new()).unwrap();
+		self.channel_monitor = Some(ChannelMonitor::new(self.local_keys.clone(),
+		                                          &self.shutdown_pubkey, self.our_to_self_delay,
+		                                          &self.destination_script, (funding_txo, funding_txo_script),
+		                                          &their_pubkeys.htlc_basepoint, &their_pubkeys.delayed_payment_basepoint,
+		                                          self.their_to_self_delay, funding_redeemscript, self.channel_value_satoshis,
+		                                          self.get_commitment_transaction_number_obscure_factor(),
+		                                          self.logger.clone()));
+
+		self.channel_monitor.as_mut().unwrap().provide_latest_remote_commitment_tx_info(&remote_initial_commitment_tx, Vec::new(), self.cur_remote_commitment_transaction_number, self.their_cur_commitment_point.unwrap());
+		self.channel_monitor.as_mut().unwrap().provide_latest_local_commitment_tx_info(local_initial_commitment_tx, local_keys, self.feerate_per_kw, Vec::new()).unwrap();
```

Comment:
710520b

Can't we just store what we need for broadcast of latest local state and drop ChannelMonitor out of Channel ? States will be only provided with new ChannelMonitorUpdates interfaces? (maybe need to cache basic info though). Also getting the minimum operation state from the Channel at anytime would let you spawn new ChannelMonitor at any-time in channel lifecycle (can we do this right now?)

Reply:
> Can't we just store what we need for broadcast of latest local state and drop ChannelMonitor out of Channel ?

Yep, thats the next step! I don't have a commit for it yet cause it requires some refactoring of ChannelMonitor and I didn't want to step on your toes *that* much :p.

> States will be only provided with new ChannelMonitorUpdates interfaces? (maybe need to cache basic info though). Also getting the minimum operation state from the Channel at anytime would let you spawn new ChannelMonitor at any-time in channel lifecycle (can we do this right now?)

Not sure what you meant by the first part, but, at least in my head, there would be no way to build a ChannelMonitor from a Channel after initialization - the docs for ManyChannelMonitor and ChannelMonitorUpdateErr always require a local store of ChannelMonitors and I don't know that there is much use in trying to get around that.

Reply:
I think we are in-sync on this, let do this after my ChannelMonitor refactor.